File: CS/MasterDetail/DataClasses.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Xml.Serialization;
using System.Collections.Generic;
using System.Collections;
using System.Linq;

namespace MasterDetail {

    public class Products {
        public int ProductID { get; set; }
        public string ProductName { get; set; }
        public int CategoryID { get; set; }
        public decimal UnitPrice { get; set; }
    }

    [XmlRoot("NewDataSet")]
    public class ProductsData : List<Products> {
        public static ProductsData DataSource {
            get {
                XmlSerializer s = new XmlSerializer(typeof(ProductsData));
                System.IO.Stream assemblyGetManifestResourceStream = typeof(ProductsData).Assembly.GetManifestResourceStream(typeof(ProductsData).Namespace + ".Data.Products.xml");
                return (ProductsData)s.Deserialize(assemblyGetManifestResourceStream);
            }
        }

        public object GetProductsByCategory(int categoryID) {
            return from product in this
                   where product.CategoryID == categoryID
                   select product;
        }
    }


    public class Categories {
        public int CategoryID { get; set; }
        public string CategoryName { get; set; }
        public string Description { get; set; }        
        public object Products { get; set; }        
    }

    [XmlRoot("NorthwindDataSet")]
    public class CategoriesData : List<Categories> {
        private static List<Categories> ReadCategories() {
            XmlSerializer s = new XmlSerializer(typeof(CategoriesData));
            System.IO.Stream assemblyGetManifestResourceStream = typeof(CategoriesData).Assembly.GetManifestResourceStream(typeof(CategoriesData).Namespace + ".Data.Categories.xml");
            return (List<Categories>)s.Deserialize(assemblyGetManifestResourceStream);
        }
        public static List<Categories> DataSource {
            get {
                List<Categories> categories = ReadCategories();
                ProductsData products = ProductsData.DataSource;
                foreach (Categories category in categories)
                    category.Products = products.GetProductsByCategory(category.CategoryID);
                return categories;
            }
        }
    }



}

File: VB/MasterDetail/DataClasses.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Net
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Documents
Imports System.Windows.Ink
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Animation
Imports System.Windows.Shapes
Imports System.Xml.Serialization
Imports System.Collections.Generic
Imports System.Collections
Imports System.Linq

Namespace MasterDetail

	Public Class Products
		Private privateProductID As Integer
		Public Property ProductID() As Integer
			Get
				Return privateProductID
			End Get
			Set(ByVal value As Integer)
				privateProductID = value
			End Set
		End Property
		Private privateProductName As String
		Public Property ProductName() As String
			Get
				Return privateProductName
			End Get
			Set(ByVal value As String)
				privateProductName = value
			End Set
		End Property
		Private privateCategoryID As Integer
		Public Property CategoryID() As Integer
			Get
				Return privateCategoryID
			End Get
			Set(ByVal value As Integer)
				privateCategoryID = value
			End Set
		End Property
		Private privateUnitPrice As Decimal
		Public Property UnitPrice() As Decimal
			Get
				Return privateUnitPrice
			End Get
			Set(ByVal value As Decimal)
				privateUnitPrice = value
			End Set
		End Property
	End Class

	<XmlRoot("NewDataSet")> _
	Public Class ProductsData
		Inherits List(Of Products)
		Public Shared ReadOnly Property DataSource() As ProductsData
			Get
				Dim s As New XmlSerializer(GetType(ProductsData))
				Dim assemblyGetManifestResourceStream As System.IO.Stream = GetType(ProductsData).Assembly.GetManifestResourceStream(GetType(ProductsData).Namespace + ".Data.Products.xml")
				Return CType(s.Deserialize(assemblyGetManifestResourceStream), ProductsData)
			End Get
		End Property

		Public Function GetProductsByCategory(ByVal categoryID As Integer) As Object
			Return _
				From product In Me _
				Where product.CategoryID = categoryID _
				Select product
		End Function
	End Class


	Public Class Categories
		Private privateCategoryID As Integer
		Public Property CategoryID() As Integer
			Get
				Return privateCategoryID
			End Get
			Set(ByVal value As Integer)
				privateCategoryID = value
			End Set
		End Property
		Private privateCategoryName As String
		Public Property CategoryName() As String
			Get
				Return privateCategoryName
			End Get
			Set(ByVal value As String)
				privateCategoryName = value
			End Set
		End Property
		Private privateDescription As String
		Public Property Description() As String
			Get
				Return privateDescription
			End Get
			Set(ByVal value As String)
				privateDescription = value
			End Set
		End Property
		Private privateProducts As Object
		Public Property Products() As Object
			Get
				Return privateProducts
			End Get
			Set(ByVal value As Object)
				privateProducts = value
			End Set
		End Property
	End Class

	<XmlRoot("NorthwindDataSet")> _
	Public Class CategoriesData
		Inherits List(Of Categories)
		Private Shared Function ReadCategories() As List(Of Categories)
			Dim s As New XmlSerializer(GetType(CategoriesData))
			Dim assemblyGetManifestResourceStream As System.IO.Stream = GetType(CategoriesData).Assembly.GetManifestResourceStream(GetType(CategoriesData).Namespace + ".Data.Categories.xml")
			Return CType(s.Deserialize(assemblyGetManifestResourceStream), List(Of Categories))
		End Function
		Public Shared ReadOnly Property DataSource() As List(Of Categories)
			Get
				Dim categories As List(Of Categories) = ReadCategories()
				Dim products As ProductsData = ProductsData.DataSource
				For Each category As Categories In categories
					category.Products = products.GetProductsByCategory(category.CategoryID)
				Next category
				Return categories
			End Get
		End Property
	End Class



End Namespace
